src/browser/api/global_hotkey.ts

```diff
@@ -0,0 +1,153 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Identity } from '../../shapes';
+import SubscriptionManager from '../subscription_manager';
+import { EventEmitter } from 'events';
+import ofEvents from '../of_events';
+import * as log from '../log';
+import route from '../../common/route';
+
+//require here because of missing definitions.
+const { globalShortcut } = require('electron');
+
+const subscriptionManager = new SubscriptionManager();
+const hotkeyOwnershipMap: Map<string, string> = new Map();
+const subCount: Map<string, number> = new Map();
+const emitter = new EventEmitter();
+
+const eventNames = {
+    REGISTERED: 'registered',
+    UNREGISTERED: 'unregistered'
+};
+
+//index.js will use these on demand
+export const reservedHotKeys: Array<string> = [
+    'CommandOrControl+0',
+    'CommandOrControl+=',
+    'CommandOrControl+Plus',
+    'CommandOrControl+-',
+    'CommandOrControl+_',
+    'CommandOrControl+Shift+I',
+    'F5',
+    'CommandOrControl+R',
+    'Shift+F5',
+    'CommandOrControl+Shift+R'
+];
+
+export class HotKeyError extends Error {
+    constructor(accelerator: string, reason: string) {
+        super(`Failed to register Hotkey: ${accelerator}, ${reason}`);
+    }
+}
+
+//want to avoid closing over any variables during the register phase.
+function constructEmit(hotkey: string): () => void {
+    return () => {
+        emitter.emit(hotkey);
+    };
+}
+
+//want to avoid closing over any variables during the register phase.
+function constructUnregister(identity: Identity, hotkey: string, listener: Function): () => void {
+    return () => {
+        emitter.removeListener(hotkey, listener);
+        if (emitter.listenerCount(hotkey) < 1) {
+            GlobalHotkey.unregister(identity, hotkey);
+        }
+    };
+}
+
+function applyRegistration(identity: Identity, hotkey: string, listener: Function): void {
+    emitter.on(hotkey, listener);
+    //make sure that if the registered context is destroyed we will unregister the hotkey
+    subscriptionManager.registerSubscription(constructUnregister(identity, hotkey, listener), identity, hotkey);
+}
+
+//here we will check if the subscription is a valid one.
+function validateRegistration(identity: Identity, hotkey: string): Error | undefined {
+    const ownerUuid = hotkeyOwnershipMap.get(hotkey);
+    // already allowed this hotkey for this uuid, return early
+    if (ownerUuid && ownerUuid === identity.uuid) {
+        return;
+    } else if (reservedHotKeys.indexOf(hotkey) > -1) {
+        return new HotKeyError(hotkey, 'is reserved');
+    } else {
+        if (globalShortcut.isRegistered(hotkey)) {
+            return new HotKeyError(hotkey, 'already registered');
+        }
+    }
+}
+
+export module GlobalHotkey {
+
+    export function register(identity: Identity, hotkey: string, listener: Function): void {
+        const validationError = validateRegistration(identity, hotkey);
```

Comment:
Why don't we throw directly in validateRegistration rather than not throw if we return undefined?

Reply:
That is a good point, making the change.